runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1536 @@
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "j9.h"
+#include "jclprots.h"
+#include "j9cp.h"
+#include "j9protos.h"
+#include "ut_j9jcl.h"
+#include "j9port.h"
+#include "jclglob.h"
+#include "jcl_internal.h"
+#include "util_api.h"
+#include "j9vmconstantpool.h"
+#include "ObjectAccessBarrierAPI.hpp"
+#include "objhelp.h"
+
+#include <string.h>
+#include <assert.h>
+
+#include "VMHelpers.hpp"
+
+extern "C" {
+
+/* Constants mapped from java.lang.invoke.MethodHandleNatives$Constants */
+#define MN_IS_METHOD		0x00010000
+#define MN_IS_CONSTRUCTOR	0x00020000
+#define MN_IS_FIELD			0x00040000
+#define MN_IS_TYPE			0x00080000
+#define MN_CALLER_SENSITIVE	0x00100000
+
+#define MN_REFERENCE_KIND_SHIFT	24
+#define MN_REFERENCE_KIND_MASK	0xF
+
+#define MN_SEARCH_SUPERCLASSES	0x00100000
+#define MN_SEARCH_INTERFACES	0x00200000
+
+/* Private MemberName object init helper 
+ *
+ * Set the MemberName fields based on the refObject given:
+ * For j.l.reflect.Field:
+ *		find JNIFieldID for refObject, create j.l.String for name and signature and store in MN.name/type fields.
+ *		set vmindex to the fieldID pointer and target to the J9ROMFieldShape struct.
+ *		set MN.clazz to declaring class in the fieldID struct.
+ * For j.l.reflect.Method or j.l.reflect.Constructor:
+ *		find JNIMethodID, set vmindex to the methodID pointer and target to the J9Method struct.
+ *		set MN.clazz to the refObject's declaring class.
+ *
+ * Then for both, compute the MN.flags using access flags and invocation type based on the JNI-id.
+ *
+ * Throw an IllegalArgumentException if the refObject is not a Field/Method/Constructor
+ */
+void
+initImpl(J9VMThread *currentThread, j9object_t membernameObject, j9object_t refObject)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
```

Comment:
It's a good habit to mark these pointers as `const`
```suggestion
	const J9JavaVM *vm = currentThread->javaVM;
	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
```